lib/attache/download.rb

```diff
@@ -97,11 +98,14 @@ def make_thumbnail_for(file, geometry, extension)
     def get_first_result_present_async(lambdas)
       return if lambdas.empty? # queue.pop will never happen
       queue = Queue.new
-      threads = lambdas.collect { |code| Thread.new { queue << code.call } }
-      until result = queue.pop do
+      threads = lambdas.shuffle.collect { |code| Thread.new { queue << [Thread.current, code.call] } }
+      until (item = queue.pop).last do
+        thread, _ = item
+        thread.join # we could be popping `queue` before thread exited
```

Comment:
@winston note that this `join` fixes the problem of attache server hanging
- in the `Thread.new` code above, we do `queue << ...` and then the thread exits
- but the processing of `queue.pop` could happen *before* the thread exits
- thus, `break unless threads.any?(&:alive?)` incorrectly account for the soon-to-be-exited thread
- the fix is to explicitly `join` that thread, before we check `threads.any?(&:alive?)`